ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java

```diff
@@ -233,42 +211,50 @@ protected ScheduledExecutorService executor() {
      * @param windowEnd the end time in ms of the current window
      */
     private void processHeartbeats(final long windowStart, final long windowEnd) {
+
+      final Map<HostInfo, HostStatus> copyOnWrite = new HashMap<>(immutableHostsStatus.get());
       // No heartbeats received -> mark all hosts as dead
       if (receivedHeartbeats.isEmpty()) {
-        hostsStatus.forEach((host, status) -> {
-          if (!host.equals(localHostString)) {
+        copyOnWrite.forEach((host, status) -> {
+          if (!host.equals(localHostInfo)) {
             status.setHostAlive(false);
           }
         });
+        immutableHostsStatus.set(new ImmutableMap.Builder<HostInfo, HostStatus>()
+                                     .putAll(copyOnWrite).build());
```

Comment:
we're copying the map twice here (once on L215 and once here). Maybe in this case it makes more sense to use `Collections.unmodifiableMap(copyOnWrite)` to enforce the immutability instead of actually making a new copy if this is performance critical code (applies generally across the code base here)

thought note, might be premature optimization given that the map is probably pretty small...

Reply:
It is a very small map, as many entries as Ksql servers in the cluster.